### PR TITLE
tikv-ctl: tombstone supports multi regions.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

**Comment by siddontang** [comment]
Seem this is not suitable for the case that 2 nodes are fine in 4 nodes.

**Comment by hicqu** [comment]
It's for **tombstone** a region, not **unsafe-recover**, which will be supported later.

**Comment by hicqu** [comment]
@overvenus @BusyJay @siddontang PTAL thanks.

File: src/server/debug.rs [hunk 1]
```diff
@@ -228,48 +228,67 @@ impl Debugger {
...
+
+            for region in regions {
+                let id = region.get_id();
+                if let Err(e) = fix_one_region(region) {
+                    ret.push((id, e));
+                }
+            }
+        }
+
+        if ret.is_empty() {
```
Comment (siddontang) [inline src/server/debug.rs]: not empty?

Comment (hicqu) [reply src/server/debug.rs]: `ret.is_empty()` is true means error occurs when we update rocksdb into write batch. In this case, we shouldn't write anything into disk. So I think it's correct.

**Comment by hicqu** [comment]
@BusyJay @overvenus PTAL thanks.

File: src/server/debug.rs [hunk 2]
```diff
@@ -228,48 +228,67 @@ impl Debugger {
...
-            .get_peers()
-            .iter()
-            .find(|p| p.get_store_id() == store_id)
-            .map_or(true, |p| p.get_id() != peer_id);
-
-        if new_conf_ver > old_conf_ver && scheduled {
-            let wb = WriteBatch::new();
-            // Here we can keep the other metas as original.
-            box_try!(write_peer_state(db, &wb, &old_region, PeerState::Tombstone));
+        let wb = WriteBatch::new();
```
Comment (BusyJay) [inline src/server/debug.rs]: It's always empty.

File: src/server/debug.rs [hunk 3]
```diff
@@ -228,48 +228,67 @@ impl Debugger {
         Ok(())
     }
 
-    /// Set a region to tombstone by manual, and apply other status(such as
+    /// Set regions to tombstone by manual, and apply other status(such as
     /// peers, version, and key range) from `region` which comes from PD normally.
-    pub fn set_region_tombstone(&self, id: u64, region: Region) -> Result<()> {
+    pub fn set_region_tombstone(&self, regions: Vec<Region>) -> Result<Vec<(u64, Error)>> {
```
Comment (BusyJay) [inline src/server/debug.rs]: Any test to cover this?

File: src/server/debug.rs (hunk 1 above)
Comment (siddontang) [reply src/server/debug.rs]: but here your logic is if empty, then write to RocksDB.

Btw, don't use this for the error assumption, if you meet error, handle the error.

**Comment by hicqu** [comment]
@siddontang @BusyJay PTAL again, thanks.

File: src/server/debug.rs [hunk 4]
```diff
@@ -228,48 +228,69 @@ impl Debugger {
...
+                let scheduled = region
+                    .get_peers()
+                    .iter()
+                    .find(|p| p.get_store_id() == store_id)
+                    .map_or(true, |p| p.get_id() != peer_id);
+                if !scheduled {
+                    return Err(box_err!("The peer is still in target peers"));
+                }
+
+                region_state.set_state(PeerState::Tombstone);
```
Comment (BusyJay) [inline src/server/debug.rs]: Why not use `write_peer_state`?

File: src/bin/tikv-ctl.rs [hunk 5]
```diff
@@ -563,9 +569,16 @@ impl DebugExecutor for Debugger {
         println!("success!");
     }
 
-    fn set_region_tombstone(&self, region_id: u64, region: Region) {
-        self.set_region_tombstone(region_id, region)
-            .unwrap_or_else(|e| perror_and_exit("Debugger::set_region_tombstone", e))
+    fn set_region_tombstone(&self, regions: Vec<Region>) {
+        let ret = self.set_region_tombstone(regions)
+            .unwrap_or_else(|e| perror_and_exit("Debugger::set_region_tombstone", e));
+        if !ret.is_empty() {
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: I think it's empty means no error happens.

File: src/server/debug.rs (hunk 4 above)
Comment (hicqu) [reply src/server/debug.rs]: call `write_peer_state` is danguerous after we support `region merge`. I think that shouldn't be treated as a general function.

Comment (BusyJay) [reply src/server/debug.rs]: How? `write_peer_state` should not be more dangerous than just put a state manually as the dangerous part are the same, which is overriding a meta key.

Comment (hicqu) [reply src/server/debug.rs]: `write_peer_state` will clean `merge_state` in `RegionLocalState`. So it will corrupt data if the region is merging.

Comment (BusyJay) [reply src/server/debug.rs]: Does it really matter? If it matters, should it be checked before marking it as tombstone? I don't think it's reason not to use `write_peer_state`.

**Comment by hicqu** [comment]
@BusyJay PTAL thanks.

**Comment by hicqu** [comment]
ping @BusyJay

File: src/server/debug.rs [hunk 6]
```diff
@@ -228,48 +228,67 @@ impl Debugger {
...
-
-        if new_conf_ver > old_conf_ver && scheduled {
-            let wb = WriteBatch::new();
-            // Here we can keep the other metas as original.
-            box_try!(write_peer_state(db, &wb, &old_region, PeerState::Tombstone));
+        let wb = WriteBatch::new();
+        let mut ret = Vec::with_capacity(regions.len());
+
+        {
+            let fix_one_region = |region: Region| {
```
Comment (BusyJay) [inline src/server/debug.rs]: This looks ugly. You can either write it in the for loop or define a function.

File: src/server/debug.rs [hunk 7]
```diff
@@ -780,6 +808,96 @@ mod tests {
...
+            target_region_1.clone(),
+            target_region_2.clone(),
+            target_region_3,
+        ];
+        let errors = debugger.set_region_tombstone(target_regions).unwrap();
+        assert_eq!(errors.len(), 1);
+        assert_eq!(errors[0].0, 3);
+
+        // After set_region_tombstone success, all region should be adjusted.
+        let target_regions = vec![target_region_1, target_region_2];
```
Comment (BusyJay) [inline src/server/debug.rs]: Should check if the regions are changed before next try.

File: src/server/debug.rs [hunk 8]
```diff
@@ -780,6 +808,96 @@ mod tests {
...
 
+    #[test]
+    fn test_tombstone_regions() {
+        let debugger = new_debugger();
+        debugger.set_store_id(11);
+        let engine = &debugger.engines.kv_engine;
+        let cf_raft = engine.cf_handle(CF_RAFT).unwrap();
+
+        // region 1 with peers at stores 11, 12, 13.
+        let mut region = Region::new();
```
Comment (BusyJay) [inline src/server/debug.rs]: Code repeats several times, extract as a function.

File: src/server/debug.rs [hunk 9]
```diff
@@ -228,48 +228,67 @@ impl Debugger {
...
+        let wb = WriteBatch::new();
+        let mut ret = Vec::with_capacity(regions.len());
+
+        {
+            let fix_one_region = |region: Region| {
+                let id = region.get_id();
+                let key = keys::region_state_key(id);
+                let store_id = self.get_store_id()?;
+
+                let region_state = box_try!(db.get_msg_cf::<RegionLocalState>(CF_RAFT, &key))
```
Comment (BusyJay) [inline src/server/debug.rs]: If it's tombstone already, can return early here.

**Comment by hicqu** [comment]
@BusyJay PTAL thanks.

**Comment by BusyJay** [comment]
LGTM

File: src/server/debug.rs [hunk 10]
```diff
@@ -228,48 +228,27 @@ impl Debugger {
...
+        let wb = WriteBatch::new();
 
-        let old_region = box_try!(db.get_msg_cf::<RegionLocalState>(CF_RAFT, &key))
-            .ok_or_else(|| Error::Other("Not a valid region".into()))
-            .map(|mut region_local_state| region_local_state.take_region())?;
+        let mut errors = Vec::with_capacity(regions.len());
+        for region in regions {
+            let region_id = region.get_id();
+            if let Err(e) = set_region_tombstone(db.as_ref(), store_id, region, &wb) {
+                errors.push((region_id, e));
```
Comment (siddontang) [inline src/server/debug.rs]: if we meet error, do we need to return directly?

Comment (hicqu) [reply src/server/debug.rs]: I think collect all errors for all regions is better. Just like if we run
```bash
find . -name *.rs -exec ls {} \;
```
if `find` meets permission denyed for some sub dirs, it won't exit immediately but keep listing others.

**Pull request merged**